tests/mesh/boundary_info.C

```diff
@@ -474,6 +478,117 @@ public:
   }
 #endif // LIBMESH_ENABLE_DIRICHLET
 
+#ifdef LIBMESH_ENABLE_AMR
+  void testBoundaryOnChildrenElements()
+  {
+    LOG_UNIT_TEST;
+
+    std::unique_ptr<Mesh> mesh;
```

Comment:
Just put the Mesh on the stack here?

But that's a quibble; this looks like an excellent first unit test.  Not yet *sufficient*, given the failure cases above, but a good start.